### DISCUSSION TITLE
Help me understand the responsibilities when creating a dialog with a complex model.

### DISCUSSION THREAD

**Discussion by uxmal** [AvaloniaUI/Avalonia#10152] [Q&A]
I'm coming from a Windows Forms background, so I'm struggling with preconceived notions that may no longen apply in the MVVM world. I'm writing a dialog view, with a complex model, and a view model which adapts the model and exposing properties for the dialog view to bind to. I'm trying to understand how to create these objects in the "canonical" or "approved" way. For concreteness, assume there is a class `EngineModel`, a class `EngineViewModel`, and a dialog view implemented in the class `EngineDialog`.  Furthermore, the main window has a `MainViewMode`, which exposes an `EngineViewModel` property. 

In the main view, I want react to a button click by showing the `EngineDialog`. After creating the dialog, should I be setting its `DataContext` property in the event handler:
```C#
private async void btnShowEngineDialog_Click(object sender, RoutedEventArgs e) {
    var dlg = new ShowEngineDialog();
    dlg.DataContext = ((MainViewModel)this.DataContext).EngineViewModel;
    await dlg.ShowDialog(this);
}
```
Or should the view model be passed in in the dialog's constructor:
```C#
private async void btnShowEngineDialog_Click(object sender, RoutedEventArgs e) {
    var engineVM = (EngineViewModel)((MainViewModel)this.DataContext).EngineViewModel;
    var dlg = new ShowEngineDialog(engineVM);
    await dlg.ShowDialog(this);
}
```
Either way, there is an awful lot of casts going on which I was hoping to avoid. What is the canonical way to pass a model or view model to a dialog?

**Comment by maxkatz6** [comment]
> I'm trying to understand how to create these objects in the "canonical" or "approved" way

I don't think there is a single canonical way.

> In the main view, I want react to a button click by showing the EngineDialog. After creating the dialog, should I be setting its DataContext property in the event handler:

In this context, you can pass it in either way. There won't be much difference. 
Only detail that if you avalonia previewer won't be able to create instance of ShowEngineDialog to preview it in the IDE, as public ctor won't exist. But IIRC you can set `Design.PreviewWith` property to some pre-created design instance.

**Reply by maxkatz6** [reply]
> Either way, there is an awful lot of casts going on which I was hoping to avoid

I am not sure why do you need to cast EngineViewModel. Isn't this property already of EngineViewModel type in your view model? Or is there a reason to not be?

**Reply by maxkatz6** [reply]
Also, probably more common would be to have some kind of integration of dialogs and windows into a view model.
Something like IApplicationController or IWindowsService interface with "ShowDialog(ViewModel)" method, and avalonia implementation which would pattern match view model type to a factory of your dialog. Having this approach, you won't have any dependencies on view later in your view model.
Or alternative and simplified, but still similar approach, with something like ReactiveUI Interactions - https://www.reactiveui.net/docs/handbook/interactions/
Where you basically have a settable public handler in the ViewModel layer, and define its implementation from the View layer.

Both of these approaches are used to make it possible to create windows and dialogs directly from the ViewModel layer, while still having View implementation isolated.

**Reply by MarkusKgit** [reply]
> Or alternative and simplified, but still similar approach, with something like ReactiveUI Interactions - https://www.reactiveui.net/docs/handbook/interactions/ Where you basically have a settable public handler in the ViewModel layer, and define its implementation from the View layer.

Or the Avalonia adopted example ;-) https://docs.avaloniaui.net/tutorials/music-store-app/opening-a-dialog
